lxd/cluster/heartbeat.go

```diff
@@ -73,46 +224,24 @@ func Heartbeat(gateway *Gateway, cluster *db.Cluster) (task.Func, task.Schedule)
 			return
 		}
 
-		heartbeats := map[int64]bool{}
-		heartbeatsLock := sync.Mutex{}
-		heartbeatsWg := sync.WaitGroup{}
+		// Cumulative set of node states (will be written back to database once done).
+		hbState := &APIHeartbeat{}
 
-		sendHeartbeat := func(id int64, address string, delay bool) {
-			defer heartbeatsWg.Done()
-
-			if delay {
-				// Spread in time by waiting up to 3s less than the interval
-				time.Sleep(time.Duration(rand.Intn((heartbeatInterval*1000)-3000)) * time.Millisecond)
-			}
-			logger.Debugf("Sending heartbeat to %s", address)
-
-			err := heartbeatNode(ctx, address, gateway.cert, raftNodes)
-			if err == nil {
-				heartbeatsLock.Lock()
-				heartbeats[id] = true
-				heartbeatsLock.Unlock()
-				logger.Debugf("Successful heartbeat for %s", address)
-			} else {
-				logger.Debugf("Failed heartbeat for %s: %v", address, err)
-			}
+		// If this leader node hasn't sent a heartbeat recently, then its node state records
+		// are likely out of date, this can happen when a node becomes a leader.
+		// Send stale set to all nodes in database to get a fresh set of active nodes.
+		if lastLeaderHeartbeat.IsZero() || time.Since(*lastLeaderHeartbeat) >= offlineThreshold {
+			logger.Warnf("Leader has not initiated heartbeat since '%v', doing initial heartbeat round", *lastLeaderHeartbeat)
+			hbState.update(false, raftNodes, allNodes, offlineThreshold)
+			hbState.send(ctx, gateway.cert, localAddress, allNodes, false)
+			logger.Debugf("Completed initial heartbeat round")
 		}
 
-		for _, node := range nodes {
-			// Special case the local node
-			if node.Address == nodeAddress {
-				heartbeatsLock.Lock()
-				heartbeats[node.ID] = true
-				heartbeatsLock.Unlock()
-				continue
-			}
-
-			// Parallelize the rest
-			heartbeatsWg.Add(1)
-			go sendHeartbeat(node.ID, node.Address, true)
-		}
-		heartbeatsWg.Wait()
+		// We have the latest set of node states, lets send that state set to all nodes.
+		hbState.update(true, raftNodes, allNodes, offlineThreshold)
+		hbState.send(ctx, gateway.cert, localAddress, allNodes, true)
```

Comment:
I also wouldn't delay this one if dealing with an initial round, we want consistency as fast as possible.

Reply:
@stgraber ok